File: scripts/bindings/ggplot_positional_temporal_break_functions.py
"""FIX-GG04 vector label functions through actual Python authoring and publication."""
import json,sys
from pathlib import Path
ROOT=Path(__file__).resolve().parents[2]
sys.path[:0]=[str(ROOT/'packages/python'),str(Path(sys.argv[1]).resolve())]
import finstack_chart as c
out=Path(sys.argv[2]);out.mkdir(parents=True,exist_ok=True)
registry=c.ExtensionRegistry.example();records=[]
output=c.Output((ROOT/'fixtures/capability/fonts/NotoSans-Regular.ttf').read_bytes());options=c.export_options(640,360).dpi(96).basis('current')
replacements_only='--replacements-only' in sys.argv[3:]
zero='--zero-range' in sys.argv[3:]
overrides='--overrides' in sys.argv[3:]
cases=json.loads((ROOT/('fixtures/parity/ggplot2/positional-temporal-break-zero.json' if zero else 'fixtures/parity/ggplot2/positional-temporal-break-overrides.json' if overrides else 'fixtures/parity/ggplot2/positional-temporal-break-functions.json')).read_text())['cases']
units=[('s','Seconds',1),('ms','Milliseconds',1000),('us','Microseconds',1000000),('ns','Nanoseconds',1000000000)]
def zone(name):
 return 'Utc' if name=='UTC' else {'Local':{'version':1,'zone':name,'revision':'1','tzdata':'explicit-US-2024','coverage':{'start':'1704067200000','end':'1735689600000'},'initial_offset_seconds':-18000,'transitions':[{'at_millis':'1710054000000','offset_seconds':-14400},{'at_millis':'1730613600000','offset_seconds':-18000}]}}
def data_for(t,unit,multiplier):
 values=t['inputs'];factor=multiplier*(86400 if t['kind']=='date' else 1)
 return c.Data.columns({'x':c.timestamps([int(v or 0)*factor for v in values],unit,t['zone']).validity([v is not None for v in values])},keys=list(range(100,100+len(values))))
def layer():return c.points().name('marks')
def build(t,d,variant,multiplier):
 date=t['kind']=='date';factor=multiplier*(86400 if date else 3600);origin=t['epoch']*multiplier
 scale=c.scale_date() if date else c.scale_utc() if t['zone']=='UTC' else c.scale_calendar({'domain':[str(origin),str(origin+3*factor)] if t['limits']=='full' else [],'unit':variant,'zone':zone(t['zone']),'range':[{'kind':'Number','value':0},{'kind':'Number','value':1}],'factory':{'kind':'Value','gamma':None},'clamp':False,'unknown':{'kind':'Missing'}})
 if t['limits']=='full' and t['zone']=='UTC':scale=scale.time_domain(origin,origin+3*factor)
 axis=c.x_axis().scale(scale).range(100.,540.).guide_geometry({'labels':'Preserve'}).breaks_function({'operation':{'id':'example.breaks_'+t['signature'],'version':'1'},'parameters':t['mode']}).tick_arguments({'count':3 if t['count']=='three' else 0 if t['count']=='zero' else None})
 if t['label_mode']=='indexed':axis=axis.tick_format({'Registered':{'operation':{'id':'example.scale_labels','version':'1'},'parameters':'indexed'}})
 if t.get('control') in ('width','both'):axis=axis.tick_arguments({'width':'1 day' if date else '1 hour'})
 if t.get('control') in ('format','both'):axis=axis.tick_format({'GgplotTime':{'pattern':'%Y-%m-%d' if date else '%H:%M','locale':None}})
 if zero:axis=axis.expansion({'mult':[0.,0.],'add':[0.,0.]})
 return c.plot(d).with_registry(registry).profile('Ggplot2_4_0_3').aes(c.aes().x({'field':'x','origin':str(origin)}).y(1.)).layer(layer()).x_axis(axis).y_axis(c.y_axis().visible(False)).build()
def check(t,frame):
 ticks=next(g for g in frame.guides()['guides'] if g['spec']['side']=='Bottom')['ticks']
 if all(isinstance(v,(float,int)) for v in t['result']['range']):
  expected=[label or '' for v,label in zip(t['result']['breaks'],t['result']['labels']) if isinstance(v,(float,int))]
  assert [tick['label'] for tick in ticks]==expected,(t,ticks,expected)
 return {'ticks':ticks}
def sample(t):
 return t['population']=='spaced' and t['limits']=='none' and t['signature']=='n' and t['count']=='three' and t['mode']=='mixed' and t['label_mode']=='automatic' and (t['kind']=='date' or t['epoch']==1710046800 or t['zone']=='America/New_York')
for index,t in ([] if replacements_only else enumerate(cases)):
 for unit,variant,multiplier in units:
  owned=[]
  try:
   d=data_for(t,unit,multiplier);owned.append(d);p=build(t,d,variant,multiplier);owned.append(p);wire=p.to_json()
   restored=c.Plot.from_json(wire,registry);owned.append(restored);assert restored.to_json()==wire
   for state in ('original','edited'):
    current=restored if state=='original' else restored.edit().layer('marks',layer()).build()
    if current is not restored:owned.append(current)
    request=output.request(current,options);owned.append(request);frame=request.prepare();owned.append(frame);actual=check(t,frame);assert 'error' not in t['result'],t
    records.append({'index':index,'unit':unit,'state':state,**actual})
    if state=='original' and unit=='ms' and sample(t):
     for fmt in ('svg','pdf','png'):(out/f"sample-{index}.{fmt}").write_bytes(frame.export(fmt))
  except c.ChartError as error:
   assert 'error' in t['result'],(t,str(error));assert error.code==('CHART_SCHEMA_CONFLICT' if 'labels' in t['result']['error'] else 'CHART_VALIDATION'),(t,error.code)
   records.append({'index':index,'unit':unit,'error':error.code})
  finally:
   for obj in reversed(owned):obj.dispose()
assert len(records)==(0 if replacements_only else 400 if zero else 1280 if overrides else 27000),len(records)
def comparable(chart):
 state=chart.semantics();domains=state['layers'][0]['domains'];space=domains.get('x_space')
 if space and 'Timestamp' in space:
  origin=int(space['Timestamp']['origin']);space['Timestamp']['origin']='0'
  if domains.get('x'):
   domains['x']={key:str(origin+int(value)) for key,value in domains['x'].items()}
 return {'limits':state.get('positional_limits'),'domains':domains}
for kind,zone_name,epoch in [('date','UTC',1704067200),('datetime','UTC',1710046800),('datetime','UTC',1730606400),('datetime','America/New_York',1710046800),('datetime','America/New_York',1730606400)]:
 if overrides or zero:break
 selected={t['population']:t for t in cases if t['kind']==kind and t['zone']==zone_name and t['epoch']==epoch and t['limits']=='none' and t['signature']=='n' and t['count']=='three' and t['mode']=='domain' and t['label_mode']=='indexed'};owned=[]
 try:
  original=data_for(selected['spaced'],'ms',1000);owned.append(original);p=build(selected['spaced'],original,'Milliseconds',1000);owned.append(p);chart=p.chart();owned.append(chart);wire=p.to_json();request=output.request(p,options);owned.append(request);held=request.prepare();owned.append(held);scene=held.scene()
  for population in ('constant','missing','all_missing','empty','spaced'):
   t=selected[population];replacement=data_for(t,'ms',1000);owned.append(replacement);tx=chart.transaction();owned.append(tx);builder=tx.replace(original,replacement);owned.append(builder);update=builder.build();owned.append(update);assert 'Applied' in chart.commit(update)
   fresh=build(t,replacement,'Milliseconds',1000);owned.append(fresh);batch=fresh.chart();owned.append(batch)
   actual=comparable(chart);assert actual==comparable(batch),(kind,zone_name,population,actual,comparable(batch));assert held.scene()==scene and p.to_json()==wire
   def render(current):
    try:
     request=output.request(current,options);owned.append(request);frame=request.prepare();owned.append(frame);assert 'error' not in t['result'];return check(t,frame)
    except c.ChartError as error:
     assert 'error' in t['result'] and error.code=='CHART_SCHEMA_CONFLICT';return {'error':error.code}
   laid=render(chart);assert laid==render(batch)
   records.append({'kind':kind,'zone':zone_name,'epoch':epoch,'replacement':population,'semantics':actual,**laid})
 finally:
  for obj in reversed(owned):obj.dispose()
assert len(records)==(25 if replacements_only else 400 if zero else 1280 if overrides else 27025)
(out/'records.json').write_text(json.dumps(records,indent=2));options.dispose();output.dispose();registry.dispose()
print(f'PASS Python: {len(records)} temporal positional break states; overrides={overrides}.')

File: crates/chart-core/tests/ggplot_positional_temporal_break_functions.rs
//! FIX-GG04: temporal label callback values, metadata, precedence and primary guides.
use chart_core::{
    ChartResult, GuideId, Rect, ResourceId, Revision,
    composition::ScaleValue,
    data::TimeUnit,
    grammar::{
        CustomGuideFormatter, CustomScaleBreaks, ExtensionDescriptor, ExtensionRegistry,
        GuideLabelsInput, OperationRef, ScaleBreaksInput, ScaleBreaksOutput,
    },
    interpolate::Number,
    layout::*,
    prelude::*,
    scales::*,
    services::*,
};
use serde_json::{Value, json};
use std::sync::{Arc, Mutex};
struct Formatter(Arc<Mutex<Vec<Value>>>);
impl CustomGuideFormatter for Formatter {
    fn descriptor(&self) -> ExtensionDescriptor {
        ExtensionDescriptor::batch("test.temporal_labels", Revision::new(1), true)
    }
    fn validate(&self, _: &Value) -> ChartResult<()> {
        Ok(())
    }
    fn format_labels(&self, input: GuideLabelsInput<'_>) -> ChartResult<Vec<Option<String>>> {
        let context = input.temporal.expect("temporal context retained");
        let n = context.normalization;
        let multiplier = match n.unit {
            TimeUnit::Seconds => 1.,
            TimeUnit::Milliseconds => 1000.,
            TimeUnit::Microseconds => 1000000.,
            TimeUnit::Nanoseconds => 1000000000.,
        };
        let factor = multiplier * if n.date { 86400. } else { 1. };
        let values = input
            .values
            .iter()
            .map(|v| {
                let ScaleValue::Number(v) = v else {
                    panic!("expected origin-relative number")
                };
                encoded_number(n.origin as f64 / factor + v / factor)
            })
            .collect::<Vec<_>>();
        let zone = match context.zone {
            CalendarZone::Utc => "UTC",
            CalendarZone::Local(rules) => &rules.zone,
        };
        self.0.lock().unwrap().push(json!({"values":values,"class":if n.date {vec!["Date"]}else{vec!["POSIXct","POSIXt"]},"zone":if n.date {vec![]}else{vec![zone]},"names":input.names.unwrap_or(&[])}));
        let mut labels = (1..=input.values.len())
            .map(|i| Some(format!("{i}/{}", input.values.len())))
            .collect::<Vec<_>>();
        if input.values.is_empty() {
            labels.push(Some("/0".into()));
        }
        match input.parameters.as_str().unwrap() {
            "missing" => {
                for (i, label) in labels.iter_mut().enumerate() {
                    if i % 2 == 1 {
                        *label = None;
                    }
                }
            }
            "short" => labels.truncate(1),
            "empty" => labels.clear(),
            "indexed" => (),
            _ => unreachable!(),
        }
        Ok(labels)
    }
}
fn zone(name: &str) -> CalendarZone {
    if name == "UTC" {
        return CalendarZone::Utc;
    }
    CalendarZone::Local(Arc::new(TimeZoneRules {
        version: 1,
        zone: name.into(),
        revision: Revision::INITIAL,
        tzdata: "explicit-US-2024".into(),
        coverage: TimeBounds {
            start: 1704067200000,
            end: 1735689600000,
        },
        initial_offset_seconds: -18000,
        transitions: vec![
            TimeZoneTransition {
                at_millis: 1710054000000,
                offset_seconds: -14400,
            },
            TimeZoneTransition {
                at_millis: 1730613600000,
                offset_seconds: -18000,
            },
        ],
    }))
}
fn normalize(v: &Value) -> Value {
    match v {
        Value::Number(n) => json!(n.as_f64().unwrap()),
        Value::Array(a) => Value::Array(a.iter().map(normalize).collect()),
        Value::Object(o) => {
            Value::Object(o.iter().map(|(k, v)| (k.clone(), normalize(v))).collect())
        }
        _ => v.clone(),
    }
}
fn encoded_number(v: f64) -> Value {
    if v.is_nan() {
        Value::Null
    } else if v == f64::INFINITY {
        json!("Infinity")
    } else if v == f64::NEG_INFINITY {
        json!("-Infinity")
    } else {
        json!(v)
    }
}
struct Breaks(Arc<Mutex<Vec<Value>>>, String);
impl CustomScaleBreaks for Breaks {
    fn descriptor(&self) -> ExtensionDescriptor {
        ExtensionDescriptor::batch("test.temporal_breaks", Revision::new(1), true)
    }
    fn accepts_n(&self) -> bool {
        self.1 == "n"
    }
    fn accepts_n_breaks(&self) -> bool {
        self.1 == "n.breaks"
    }
    fn validate(&self, _: &Value) -> ChartResult<()> {
        Ok(())
    }
    fn evaluate(&self, input: ScaleBreaksInput<'_>) -> ChartResult<ScaleBreaksOutput> {
        let context = input.temporal.expect("typed temporal limits");
        let n = context.normalization;
        let multiplier = match n.unit {
            TimeUnit::Seconds => 1.,
            TimeUnit::Milliseconds => 1e3,
            TimeUnit::Microseconds => 1e6,
            TimeUnit::Nanoseconds => 1e9,
        };
        let factor = multiplier * if n.date { 86400. } else { 1. };
        let d = input
            .domain
            .iter()
            .map(|v| match v {
                ScaleKey::Number(n) => n.0,
                _ => panic!("numeric offsets"),
            })
            .collect::<Vec<_>>();
        let zone = match context.zone {
            CalendarZone::Utc => "UTC",
            CalendarZone::Local(r) => &r.zone,
        };
        assert_eq!(input.count_argument, input.count.map(|_| "n"));
        self.0.lock().unwrap().push(json!({"values":d.iter().map(|v| encoded_number(n.origin as f64/factor+v/factor)).collect::<Vec<_>>(),
            "class":if n.date {vec!["Date"]}else{vec!["POSIXct","POSIXt"]}, "zone":if n.date {vec![]}else{vec![zone]}, "names":[], "count":input.count,
            "effective":match self.1.as_str() { "n"=>json!(input.count.unwrap_or(7.)), "n.breaks"=>json!(9.), _=>Value::Null }}));
        let mode = input.parameters.as_str().unwrap();
        let (values, names) = match mode {
            "domain" | "numeric" => (Some(d), None),
            "mixed" => (
                Some(vec![
                    d[1],
                    (d[0] + d[1]) / 2.,
                    d[0],
                    d[0],
                    f64::NAN,
                    f64::INFINITY,
                    f64::NEG_INFINITY,
                ]),
                Some(
                    [
                        "last", "middle", "first", "again", "missing", "positive", "negative",
                    ]
                    .map(String::from)
                    .to_vec(),
                ),
            ),
            "empty" => (Some(vec![]), None),
            "null" => (None, None),
            _ => unreachable!(),
        };
        Ok(ScaleBreaksOutput {
            values: values.map(|v| v.into_iter().map(|v| ScaleKey::Number(Number(v))).collect()),
            names,
            temporal: (mode != "numeric" && mode != "null").then_some(n),
        })
    }
}
struct Metrics;
impl TextMeasurer for Metrics {
    fn measure(&self, r: TextRequest<'_>) -> ChartResult<TextMetrics> {
        TextMetrics::new(r.text.len() as f64 * 5., 8., 2.)
    }
}
#[test]
fn positional_temporal_breaks_match_reference() {
    assert_positional_temporal_breaks(
        include_str!("../../../fixtures/parity/ggplot2/positional-temporal-break-functions.json"),
        9000,
    );
}
#[test]
fn positional_temporal_break_precedence_matches_reference() {
    assert_positional_temporal_breaks(
        include_str!("../../../fixtures/parity/ggplot2/positional-temporal-break-overrides.json"),
        560,
    );
}
#[test]
fn positional_temporal_zero_range_bypasses_break_callbacks() {
    assert_positional_temporal_breaks(
        include_str!("../../../fixtures/parity/ggplot2/positional-temporal-break-zero.json"),
        200,
    );
}
fn assert_positional_temporal_breaks(source: &str, expected_success: usize) {
    let fixture: Value = serde_json::from_str(source).unwrap();
    let mut passed = 0;
    for (index, c) in fixture["cases"].as_array().unwrap().iter().enumerate() {
        for (unit, multiplier) in [
            (TimeUnit::Seconds, 1_i64),
            (TimeUnit::Milliseconds, 1000),
            (TimeUnit::Microseconds, 1000000),
            (TimeUnit::Nanoseconds, 1000000000),
        ] {
            let date = c["kind"] == "date";
            let origin = c["epoch"].as_i64().unwrap() * multiplier;
            let factor = multiplier * if date { 86400 } else { 3600 };
            let calls = Arc::new(Mutex::new(vec![]));
            let breaks = Arc::new(Mutex::new(vec![]));
            let mut registry = ExtensionRegistry::default();
            registry
                .register_guide_formatter(Arc::new(Formatter(calls.clone())))
                .unwrap();
            registry
                .register_scale_breaks(Arc::new(Breaks(
                    breaks.clone(),
                    c["signature"].as_str().unwrap().into(),
                )))
                .unwrap();
            let inputs = c["inputs"].as_array().unwrap();
            let stamps = inputs
                .iter()
                .map(|v| {
                    (v.as_f64().unwrap_or(0.) * multiplier as f64 * if date { 86400. } else { 1. })
                        as i64
                })
                .collect::<Vec<_>>();
            let data = Data::columns()
                .column(
                    "x",
                    timestamps(stamps, unit, c["zone"].as_str().unwrap())
                        .validity(inputs.iter().map(|v| !v.is_null()).collect()),
                )
                .column("y", vec![1.; inputs.len()])
                .build()
                .unwrap();
            let mut scale = if date {
                scale_date()
            } else if c["zone"] == "UTC" {
                scale_utc()
            } else {
                scale_calendar(TimeScaleSpec {
                    domain: if c["limits"] == "full" {
                        vec![origin, origin + 3 * factor]
                    } else {
                        vec![]
                    },
                    unit,
                    zone: zone(c["zone"].as_str().unwrap()),
                    ..Default::default()
                })
            };
            if c["limits"] == "full" && c["zone"] == "UTC" {
                scale = scale.time_domain(origin, origin + 3 * factor);
            }
            let mut axis =
                x_axis()
                    .scale(scale)
                    .range(100., 540.)
                    .guide_geometry(Some(GuideGeometry {
                        labels: Some(GuideLabelPolicy::Preserve),
                        ..Default::default()
                    }));
            axis = axis
                .breaks_function(Some(chart_core::grammar::ScaleBreaksOperation {
                    operation: OperationRef::new("test.temporal_breaks", Revision::new(1)),
                    parameters: c["mode"].clone(),
                }))
                .tick_arguments(Some(GuideTickArguments {
                    count: match c["count"].as_str().unwrap() {
                        "three" => Some(3.),
                        "zero" => Some(0.),
                        _ => None,
                    },
                    ..Default::default()
                }));
            if c["label_mode"] == "indexed" {
                axis = axis.tick_format(Some(GuideFormatter::Registered {
                    operation: OperationRef::new("test.temporal_labels", Revision::new(1)),
                    parameters: json!("indexed"),
                }));
            }
            if matches!(c["control"].as_str(), Some("width" | "both")) {
                axis = axis.tick_arguments(Some(GuideTickArguments {
                    count: None,
                    width: Some(if date { "1 day" } else { "1 hour" }.into()),
                    ..Default::default()
                }));
            }
            if matches!(c["control"].as_str(), Some("format" | "both")) {
                axis = axis.tick_format(Some(GuideFormatter::GgplotTime(Box::new(
                    GgplotTimeFormat {
                        pattern: if date { "%Y-%m-%d" } else { "%H:%M" }.into(),
                        locale: None,
                    },
                ))));
            }
            if c["control"] == "zero" {
                axis = axis.expansion(Some(GgplotExpansion {
                    mult: [0.; 2],
                    add: [0.; 2],
                }));
            }
            let result = (|| -> ChartResult<_> {
                let p = plot(data)
                    .profile(Profile::Ggplot2_4_0_3)
                    .aes(
                        aes()
                            .x(Mapping::Timestamp {
                                field: "x".into(),
                                origin,
                            })
                            .y("y"),
                    )
                    .layer(points())
                    .x_axis(axis)
                    .extensions(Arc::new(registry))
                    .build()?;
                layout(
                    p.chart()?.prepare()?,
                    &LayoutRequest::new(
                        Rect::new(0., 0., 640., 360.)?,
                        Units::LogicalPixels,
                        ResourceDescriptor {
                            id: ResourceId::new(1),
                            revision: Revision::INITIAL,
                            kind: ResourceKind::Font,
                            byte_len: 1,
                        },
                    ),
                    &Metrics,
                )
            })();
            assert_eq!(
                result.is_ok(),
                c["result"].get("error").is_none(),
                "case {index} {unit:?} {c}: {:?}",
                result.as_ref().err()
            );
            let observed = breaks.lock().unwrap();
            let expected_breaks = c["calls"].as_array().unwrap();
            assert_eq!(
                observed.is_empty(),
                expected_breaks.is_empty(),
                "break call presence {index}"
            );
            for call in observed.iter() {
                assert_eq!(
                    normalize(call),
                    normalize(&expected_breaks[0]),
                    "break call {index} {unit:?}"
                );
            }
            let actual = calls.lock().unwrap();
            let expected = c["label_calls"].as_array().unwrap();
            assert_eq!(
                actual.is_empty(),
                expected.is_empty(),
                "case {index}: {actual:?} != {expected:?}"
            );
            for call in actual.iter() {
                let expected = &expected[0];
                for key in ["class", "zone", "names"] {
                    assert_eq!(call[key], expected[key], "case {index} {key}");
                }
                let a = call["values"].as_array().unwrap();
                let b = expected["values"].as_array().unwrap();
                assert_eq!(a.len(), b.len(), "case {index}: {a:?} != {b:?}");
                for (a, b) in a.iter().zip(b) {
                    assert_eq!(
                        normalize(a),
                        normalize(b),
                        "case {index} {unit:?}: {a} != {b}"
                    );
                }
            }
            if let Ok(frame) = result {
                passed += 1;
                if c["result"]["range"]
                    .as_array()
                    .unwrap()
                    .iter()
                    .all(Value::is_number)
                {
                    let wanted = c["result"]["breaks"]
                        .as_array()
                        .unwrap()
                        .iter()
                        .zip(c["result"]["labels"].as_array().unwrap())
                        .filter(|(v, _)| v.is_number())
                        .map(|(_, v)| v.as_str().unwrap_or(""))
                        .collect::<Vec<_>>();
                    let got = frame.guides()[&GuideId::new(0)]
                        .ticks
                        .iter()
                        .map(|t| t.label.as_str())
                        .collect::<Vec<_>>();
                    assert_eq!(got, wanted, "case {index}");
                    let range = c["result"]["range"].as_array().unwrap();
                    let low = range[0].as_f64().unwrap();
                    let high = range[1].as_f64().unwrap();
                    for (tick, expected) in frame.guides()[&GuideId::new(0)].ticks.iter().zip(
                        c["result"]["breaks"]
                            .as_array()
                            .unwrap()
                            .iter()
                            .filter_map(Value::as_f64),
                    ) {
                        let factor = multiplier as f64 * if date { 86400. } else { 1. };
                        let actual = match tick.value {
                            ScaleValue::Timestamp { value, .. } => value as f64 / factor,
                            ScaleValue::Number(n) => origin as f64 / factor + n / factor,
                            _ => unreachable!(),
                        };
                        assert_eq!(actual, expected, "tick value {index} {unit:?}");
                        let position = if low == high {
                            320.
                        } else {
                            100. + 440. * (expected - low) / (high - low)
                        };
                        assert!(
                            (tick.position - position).abs() < 1e-8,
                            "tick position {index} {unit:?}: {} != {position}",
                            tick.position
                        );
                    }
                }
            }
        }
    }
    assert_eq!(passed, expected_success);
}
